# Data Exploration

In [2]:
import pandas as pd
from pathlib import Path

In [77]:
def dataset_info(df):
# Create a DataFrame to contain all the information about the dataset
    # Get the columns from the dataset
    columns = df.columns.to_list()

    # Add the columns to a dictionary
    info_dict = {'columns': columns}

    # Transform the dictionary into a DataFrame
    info_df = pd.DataFrame(info_dict)

    # Add columns to store the info
    info_df['dtypes'] = ''
    info_df['elements'] = 0
    info_df['missing'] = 0
    info_df['unique'] = 0

    # Initialise counter
    count = 0

    # Get information for all columns
    for col in columns:
        info_df.iloc[count,1] = str(df[col].dtypes)
        info_df.iloc[count,2] = df[col].count()
        info_df.iloc[count,3] = len(df) - df[col].count()
        info_df.iloc[count,4] = df[col].nunique()
        count += 1

    return info_df

## Explore the dataset
- columns
- data format
- missing data (NaN, Null, etc.)

In [3]:
# Read the CSV into a DataFrame
csv = Path('Citibike_Data/JC-202307-citibike-tripdata.csv')
trip_df = pd.read_csv(csv)

In [79]:
# Get information about the dataset
print(f"The dataset contains {len(trip_df)} rows")
dataset_info(trip_df)

The dataset contains 106608 rows


columns   dtypes  elements  missing  unique
0              ride_id   object    106608        0  106608
1        rideable_type   object    106608        0       3
2           started_at   object    106608        0  103107
3             ended_at   object    106608        0  103481
4   start_station_name   object    106601        7     103
5     start_station_id   object    106601        7     103
6     end_station_name   object    106216      392     191
7       end_station_id   object    106216      392     191
8            start_lat  float64    106608        0   34355
9            start_lng  float64    106608        0   35009
10             end_lat  float64    106486      122     250
11             end_lng  float64    106486      122     249
12       member_casual   object    106608        0       2

## Early observation
- 103 start startions but 191 end station, are some stations never used as starting points? Where are there?
- 392 end stations are missing, where the bikes stolen? Or just returned in the wild? See that only 122 end_lat/end_lng are missing, how far from a station are these bikes?
- More trips ended than started? Where they started the previous month?

## Follow up questions
- What dates are covered? Earliest/latest start and end dates?
- What are the ride types?
- How many members vs casual?

In [92]:
print(f"First start time: {min(trip_df['started_at'])}")
print(f"Last start time: {max(trip_df['started_at'])}")
print(f"First end time: {min(trip_df['ended_at'])}")
print(f"Last end time: {max(trip_df['ended_at'])}")

First start time: 2023-07-01 00:00:29
Last start time: 2023-07-31 23:59:44
First end time: 2023-07-01 00:03:56
Last end time: 2023-08-01 23:11:42


In [84]:
trip_df['rideable_type'].value_counts()

classic_bike     95055
electric_bike    11338
docked_bike        215
Name: rideable_type, dtype: int64

In [85]:
trip_df['member_casual'].value_counts()

member    73344
casual    33264
Name: member_casual, dtype: int64

## Ride duration

23.0